support for test code reuse (#320)
ocaml-sf/learn-ocaml: docs/tutorials/step-8.md

# Step 8: Reusing the grader code

This step explains how to separate the grader code, and eventually reuse it in other exercises. 

During the grading, the file **test.ml** is evaluated in an environment that contains notably: 
- **prelude.ml** and **prepare.ml** ;
- the student code isolated in a module `Code` ;
- **solution.ml** in a module `Solution` ;
- the grading modules **Introspection**, **Report** and **Test_lib**.

### Separating the grader code

It is possible to extend this environment by declaring some other user-defined modules in an optional file **depend.txt**, located in the exercise directory.

Each declaration in **depend.txt** is a single line containing the relative path of an *.ml* or *.mli* file. The order of the *.ml* declarations specifies the order in which each module is loaded in the grading environment.

> By default each dependency *foo.ml* is isolated in a module *Foo*, which can be constrained by the content of an optional signature file *foo.mli*. Furthermore, we can add an annotation `@included` at the beginning of a file *foo.ml* to denote that all the bindings of *foo.ml* is are evaluated in the toplevel environment (and not in a module *Foo*). 

Reviewer: Please use shorter lines in text files (80 columns max.).

Here, who is "we"?

Reviewer: ```suggestion
By default each dependency *foo.ml* is isolated in a module *Foo*, which can be constrained by the content of an optional signature file *foo.mli*. Furthermore, we can add an annotation `@included` at the beginning of a file *foo.ml* to denote that all the bindings of *foo.ml* are evaluated in the toplevel environment (and not in a module *Foo*). 
```

Reviewer: You should provide an example, especially to show what is exactly the syntax for annotations.

Reviewer: No that's fine.

> Note that when you edited an *.ml* or *.mli* dependency file which is not defined at the root of your exercice directory, the change will not be applied. In this case a simple shell command `$ touch test.ml` refresh the exercise.

Reviewer: ```suggestion
Dependencies that are not defined at the root of the exercise repository are ignored by the build system: therefore, if you modify them, do not forget to refresh the timestamp of `test.ml` (using `touch` for instance).
```

Author: thank you. It's updated.

### A complete example

Let's write an exercise dedicated to *Peano numbers*. Here is the structure of the exercise:

```
.
├── exercises
│   ├── index.json
│   └── lib
│   │   ├── check.ml
│   │   └── check.mli
│   ├── peano
│   │   ├── depend.txt
│   │   ├── descr.md
│   │   ├── meta.json
│   │   ├── prelude.ml
│   │   ├── prepare.ml
│   │   ├── solution.ml
│   │   ├── template.ml
│   │   ├── test.ml
│   │   └── tests
│   │   │   ├── samples.ml
│   │       ├── add.ml
│   │       └── odd_even.ml
│   ├── an-other-exercise
│   │   ├── depend.txt
│   │   │ ...
```

> The exercise **peano** follows the classical format : **prelude.ml**, **prepare.ml**, **solution.ml**, **template.ml** and **test.ml**. It comprises moreover several dependencies (**check.ml**, **samples.ml**, **add.ml** and **odd_even.ml**) which are declared as follow in **depend.txt**:

Reviewer: ```suggestion
The exercise **peano** follows the classical format : **prelude.ml**, **prepare.ml**, **solution.ml**, **template.ml** and **test.ml**. It also includes several dependencies (**check.ml**, **samples.ml**, **add.ml** and **odd_even.ml**) which are declared as follows in **depend.txt**:
```

Author: thank you. It's updated.

```ocaml
let rec add n = function
| Z -> n
| S m -> S (add n m)

let rec odd = function
| Z -> false 
| S n -> even n
and even = function
| Z -> true
| S n -> odd n
```

- **test.ml**
```ocaml
let () =
  Check.safe_set_result [ Add.test ; Odd_even.test ]
```

Note that **test.ml** is very compact because it simply combines functions defined in separated files.

- **../lib/check.ml**:
```ocaml
open Test_lib
open Report

let safe_set_result tests =
  set_result @@
  ast_sanity_check code_ast @@ fun () ->
    List.mapi (fun i test -> 
                Section ([ Text ("Question " ^ string_of_int i ^ ":") ],
                         test ())) tests
```
- **../lib/check.mli**:
```ocaml
val safe_set_result : (unit -> Report.t) list -> unit
```

- **tests/add.ml**: 
```ocaml
let test () =
  Test_lib.test_function_2_against_solution
    [%ty : peano -> peano -> peano ] "add"
    [ (Z, Z) ; (S(Z), S(S(Z))) ]
```
- **tests/odd_even.ml** :
```ocaml
let test () =
  Test_lib.test_function_1_against_solution
    [%ty : peano -> bool ] "odd"
    [ Z ; S(Z) ; S(S(Z)) ] 
  @
  Test_lib.test_function_1_against_solution
    [%ty : peano -> bool ] "even"
    [ Z ; S(Z) ; S(S(Z)) ]
```

> Remember that **Test_lib** internally requires a user-defined sampler `sample_peano : unit -> peano` to generate value of type `peano`. This sampler has to be present in the toplevel environment -- and not in a module -- in order to be found by the introspection primitives during grading. Therefore, we define this sampler in a file starting by the annotation `@included`.

Reviewer: ```suggestion
Remember that **Test_lib** internally requires a user-defined sampler `sample_peano : unit -> peano` to generate value of type `peano`. This sampler has to be present in the toplevel environment -- and not in a module -- in order to be found by the introspection primitives during grading. Therefore, we define this sampler in a file starting with the annotation `@included`.
```

Author: thank you. It's updated.

- **samples.ml**:
```ocaml

> @included

Reviewer: Using the standard syntax:
```
[@@@included]
```
would have the benefits of making this .ml file accepted by the compiler.

Author: I have switched to the standard syntax.

Note that the parsing of the annotation is hard-coded [in grading.ml](https://github.com/lsylvestre/learn-ocaml/blob/e99fbc9d9e020d9abe6d641c5534b40b6e548111/src/grader/grading.ml#L167) as a string comparison:
 ```ocaml
match String.trim (String.sub content 0 i) with 
| "[@@@included]" -> ...
```

Is this the right way to do it?

Reviewer: That's probably fine.

A cleaner way would be to inspect the module syntax tree but I am not sure if that is easy for you to get this tree at this point of the process.
